Feat/codelab basics 1 (#878)
open-wc/open-wc: packages/codelabs/src/basics/1-web-component.md

To do this efficiently, browsers have a template element. This allows us to define the structure of a piece of HTML once upfront, and efficiently clone this each time the element is rendered on the page. This is a lot faster than recreating the same HTML structure each time. Writing and cloning templates is (intentionally) pretty low level, you will usually not want to do this manually. You can read more about the basic API [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template).

For writing templates, we recommend the [lit-html](https://github.com/Polymer/lit-html) library. We will use [lit-element](https://github.com/Polymer/lit-element), which makes it easy to use it in a web component.

LitElement is written and distributed as es modules, this means we can import it using the browser's native module loader. Let's create a module script, and import `LitElement` from a CDN:

```html
<!DOCTYPE html>

<html>
  <body>
    <script type="module">
      import { LitElement } from 'https://unpkg.com/lit-element?module';

    </script>
  </body>
</html>
```

> Make sure you add type="module" to the script tag!

Reviewer: ```suggestion
Make sure you add `type="module"` to the script tag!
```

Next, we need to define our class. Instead of extending `HTMLElement`, we are now extending `LitElement`. `LitElement` extends `HTMLElement` already, so we're still writing an actual web component:

```js
class WebComponentApis extends LitElement {
  connectedCallback() {

>     // make

Reviewer: make?

Author: woops, will remove that one

    super.connectedCallback();
    console.log('lit element connected');
  }
}

customElements.define('web-component-apis', WebComponentApis);
```

```html
  <web-component-apis></web-component-apis>
```

> If you run this in the browser you should see `hello world` logged to the terminal.

Reviewer: ```suggestion
If you run this in the browser you should see `lit element connected` logged to the terminal.
```

<details>
  <summary>
    View final result
  </summary>

```html
<!DOCTYPE html>

<html>
  <body>
    <style>
      body {
        font-family: monospace;
      }

      h1 {
        color: pink;
      }

      ul {
        font-weight: bold;
        list-style-type: decimal;
      }
    </style>

    <h1>Basic Web Components APIs</h1>

    <ul>
      <li>Custom Elements</li>
      <li>Templates</li>
      <li>Shadow DOM</li>
    </ul>

    <web-component-apis></web-component-apis>

    <script type="module">
      import { LitElement, html } from 'https://unpkg.com/lit-element?module';

      class WebComponentApis extends LitElement {
        render() {
          return html`
            <h1>Basic Web Components APIs</h1>

            <ul>
              <li>Custom Elements</li>
              <li>Templates</li>
              <li>Shadow DOM</li>
            </ul>
          `;
        }
      }

      customElements.define('web-component-apis', WebComponentApis);
    </script>
  </body>
</html>
```
</details>

## Web Components in the wild
Web Components are being used in the wild by many projects and companies. Some examples:

### Github

> Github uses web component for various part of their website. They're using just the Custom Elements API, relying on global styling. They use them as a progressive enhancements, on browsers without support or when javascript is turned off, there is a fallback text that is dispalyed.

Reviewer: ```suggestion
Github uses web component for various part of their website. They're using just the Custom Elements API, relying on global styling. They use them as a progressive enhancements, on browsers without support or when javascript is turned off, there is a fallback text that is displayed.
```
